http2/src/main/java/com/blazemeter/jmeter/http2/sampler/HTTP2Connection.java

```diff
@@ -69,86 +70,91 @@ public boolean isClosed() {
-                        String v = header.getValue();
-                        v = v.replaceFirst(":\\d+$", ""); // remove any port
-                        // specification //
-                        // $NON-NLS-1$
-                        // $NON-NLS-2$
-                        requestFields.put(n, v);
-                        headerString.append(n).append(": ").append(v).append("\n");
-                    }
-                }
-            }
-            // TODO CacheManager
-        }
-
-        sampleResult.sampleStart();
-
-        // Extracts all the required cookies for that particular URL request
-        String cookieHeader = null;
-        if (cookieManager != null) {
-            cookieHeader = cookieManager.getCookieHeaderForURL(url);
-            if (cookieHeader != null) {
-                requestFields.put(HTTPConstants.HEADER_COOKIE, cookieHeader);
-                headerString.append(HTTPConstants.HEADER_COOKIE).append(": ").append(cookieHeader).append("\n");
+                     RequestBody requestBody, HTTP2SampleResult sampleResult, int timeout) throws Exception {
+        HttpFields headers = buildHeaders(url, headerManager, cookieManager);
+
+        if (requestBody != null) {
+            headers.put(HTTPConstants.HEADER_CONTENT_LENGTH, Long.toString(requestBody.getPayloadBytes().length));
+            // Check if the header manager had a content type header
+            // This allows the user to specify his own content-type for a POST request
+            String contentTypeHeader = headers.get(HTTPConstants.HEADER_CONTENT_TYPE);
+            if (contentTypeHeader == null || contentTypeHeader.isEmpty()) {
+                headers.put(HTTPConstants.HEADER_CONTENT_TYPE, HTTPConstants.APPLICATION_X_WWW_FORM_URLENCODED);
             }
+            sampleResult.setQueryString(requestBody.getPayload());
         }
 
         MetaData.Request metaData = null;
         boolean endOfStream = true;
         switch (method) {
             case "GET":
                 metaData = new MetaData.Request("GET", new HttpURI(url.toString()), HttpVersion.HTTP_2,
-                        requestFields);
+                        headers);
                 break;
             case "POST":
                 metaData = new MetaData.Request("POST", new HttpURI(url.toString()), HttpVersion.HTTP_2,
-                        requestFields);
+                        headers);
                 endOfStream = false;
                 break;
             default:
                 break;
         }
 
         HeadersFrame headersFrame = new HeadersFrame(metaData, null, endOfStream);
-        sampleResult.setRequestHeaders(headerString.toString());
-        sampleResult.setBytes(sampleResult.getBytesAsLong() + (long) headerString.length());
+        // we do this replacement and remove final char to be consistent with jmeter HTTP request sampler
+        String headersString = headers.toString().replaceAll("\r\n", "\n");
```

Comment:
@undera What do you think about this replacement?

Reply:
I have no strong opinion here. If this is really JMeter standard way - I'm ok.

Reply:
Jmeter http plugin is using as http request and response headers the version without the \r\n and without the final \r\n, but in the headers size they count includes them. We decided to do the same, and the replacement in this case is necessary since HttpFields::toString returns the \r\n version.